src/include/api.h

```diff
@@ -35,27 +35,30 @@
 #define API_SESSION_PUSH(s, h, n, dh)       \
     WT_DATA_HANDLE *__olddh = (s)->dhandle; \
     const char *__oldname = (s)->name;      \
+    ++s->api_call_counter;                  \
     (s)->dhandle = (dh);                    \
     (s)->name = (s)->lastop = #h "." #n
 #define API_SESSION_POP(s)  \
     (s)->dhandle = __olddh; \
-    (s)->name = __oldname
+    (s)->name = __oldname;  \
+    --s->api_call_counter
 
 /* Standard entry points to the API: declares/initializes local variables. */
-#define API_SESSION_INIT(s, h, n, dh)                              \
-    WT_TRACK_OP_DECL;                                              \
-    API_SESSION_PUSH(s, h, n, dh);                                 \
-    /*                                                             \
-     * No code before this line, otherwise error handling won't be \
-     * correct.                                                    \
-     */                                                            \
-    WT_ERR(WT_SESSION_CHECK_PANIC(s));                             \
-    WT_SINGLE_THREAD_CHECK_START(s);                               \
-    WT_TRACK_OP_INIT(s);                                           \
-    __wt_op_timer_start(s);                                        \
-    /* Reset wait time if this isn't an API reentry. */            \
-    if (__oldname == NULL)                                         \
-        (s)->cache_wait_us = 0;                                    \
+#define API_SESSION_INIT(s, h, n, dh)                                 \
+    WT_TRACK_OP_DECL;                                                 \
+    API_SESSION_PUSH(s, h, n, dh);                                    \
+    /*                                                                \
+     * No code before this line, otherwise error handling won't be    \
+     * correct.                                                       \
+     */                                                               \
+    WT_ERR(WT_SESSION_CHECK_PANIC(s));                                \
+    WT_SINGLE_THREAD_CHECK_START(s);                                  \
+    WT_TRACK_OP_INIT(s);                                              \
+    if (s->api_call_counter == 1 && !F_ISSET(s, WT_SESSION_INTERNAL)) \
+        __wt_op_timer_start(s);                                       \
+    /* Reset wait time if this isn't an API reentry. */               \
+    if (__oldname == NULL)                                            \
+        (s)->cache_wait_us = 0;                                       \
```

Comment:
Could you explain what this is for? Is this a tricky way of disabling the timeout for non-cursor methods? How does it ever get reset back to the configured value?

Reply:
This is the timer for cache wait time, which relates to the code that may throw WT_CACHE_FULL if cache wait time is configured:
```
        if (cache->cache_max_wait_us != 0 && session->cache_wait_us > cache->cache_max_wait_us) {
            WT_TRET(WT_CACHE_FULL);
            WT_STAT_CONN_INCR(session, cache_timed_out_ops);
        }
```
It assumes if we enter the api call the first time (name will be NULL), we need to reset the timer. I agree this is tricky and we should switch it to the same reference counter mechanism to reset it.